File: src/main/java/com/uml/generator/generator/DeploymentDiagramGenerator.java
package com.uml.generator.generator;

import com.uml.generator.analyzer.TraceAggregator;
import com.uml.generator.model.Trace;
import com.uml.generator.renderer.XmiWriter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

import java.util.*;

/**
 * Generates UML Deployment Diagrams in XMI 2.5.1 format from aggregated Jaeger
 * traces.
 * Extracts deployment information from span tags (hostname, pod, namespace,
 * etc.)
 */
public class DeploymentDiagramGenerator implements DiagramGenerator {

    private static final Logger logger = LoggerFactory.getLogger(DeploymentDiagramGenerator.class);
    private static final String UML_NAMESPACE = "http://www.eclipse.org/uml2/5.0.0/UML";

    @Override
    public String getDiagramType() {
        return "deployment";
    }

    @Override
    public String generateXmi(List<Trace> traces) {
        if (traces == null || traces.isEmpty()) {
            logger.warn("No traces provided for deployment diagram generation");
            return "";
        }

        try {
            TraceAggregator aggregator = new TraceAggregator(traces);
            XmiWriter writer = new XmiWriter();

            // Determine model name
            String modelName = "DeploymentDiagram";
            if (!traces.isEmpty() && traces.get(0).getSourceName() != null) {
                modelName = traces.get(0).getSourceName() + "_Deployment";
            }

            // Create XMI document
            Document doc = writer.createXmiDocument(modelName);
            Element model = writer.getModelElement(doc);

            // Extract deployment information from service metadata
            Map<String, Map<String, Object>> serviceMetadata = aggregator.getServiceMetadata();
            Map<String, Set<String>> nodeServices = new HashMap<>();
            Map<String, String> nodeIds = new HashMap<>();

            // Analyze metadata to extract deployment info
            for (Map.Entry<String, Map<String, Object>> entry : serviceMetadata.entrySet()) {
                String service = entry.getKey();
                Map<String, Object> metadata = entry.getValue();

                String node = extractNode(metadata);
                nodeServices.computeIfAbsent(node, k -> new HashSet<>()).add(service);
            }

            // Create nodes and deploy artifacts
            for (Map.Entry<String, Set<String>> nodeEntry : nodeServices.entrySet()) {
                String nodeName = nodeEntry.getKey();
                Set<String> services = nodeEntry.getValue();

                // Create node
                String nodeId = createNode(doc, model, nodeName);
                nodeIds.put(nodeName, nodeId);

                // Create artifacts and deployments for services on this node
                for (String service : services) {
                    String artifactId = createArtifact(doc, model, service);
                    createDeployment(doc, model, nodeName + "_deploys_" + service,
                            nodeId, artifactId);
                }
            }

            // Create communication paths between nodes
            Map<String, Set<String>> dependencies = aggregator.getServiceDependencies();
            Set<String> createdPaths = new HashSet<>();

            for (Map.Entry<String, Set<String>> entry : dependencies.entrySet()) {
                String fromService = entry.getKey();
                Set<String> toServices = entry.getValue();

                // Find nodes for services
                String fromNode = findNodeForService(fromService, nodeServices);

                for (String toService : toServices) {
                    String toNode = findNodeForService(toService, nodeServices);

                    if (fromNode != null && toNode != null && !fromNode.equals(toNode)) {
                        String pathKey = fromNode + "_to_" + toNode;
                        String pathKeyReverse = toNode + "_to_" + fromNode;

                        // Create bidirectional path only once
                        if (!createdPaths.contains(pathKey) && !createdPaths.contains(pathKeyReverse)) {
                            String fromNodeId = nodeIds.get(fromNode);
                            String toNodeId = nodeIds.get(toNode);

                            if (fromNodeId != null && toNodeId != null) {
                                createCommunicationPath(doc, model, pathKey, fromNodeId, toNodeId);
                                createdPaths.add(pathKey);
                            }
                        }
                    }
                }
            }

            int totalNodes = nodeServices.size();
            int totalServices = serviceMetadata.size();

            logger.info("Generated deployment diagram XMI with {} node(s) and {} service(s)",
                    totalNodes, totalServices);

            return writer.documentToString(doc);

        } catch (Exception e) {
            logger.error("Failed to generate deployment diagram XMI", e);
            return "";
        }
    }

    /**
     * Creates a UML Node element.
     */
    private String createNode(Document doc, Element model, String nodeName) {
        Element node = doc.createElementNS(UML_NAMESPACE, "packagedElement");
        node.setAttribute("xmi:type", "uml:Node");
        String nodeId = XmiWriter.generateUUID();
        node.setAttribute("xmi:id", nodeId);
        node.setAttribute("name", nodeName);

        model.appendChild(node);
        return nodeId;
    }

    /**
     * Creates a UML Artifact element.
     */
    private String createArtifact(Document doc, Element model, String artifactName) {
        Element artifact = doc.createElementNS(UML_NAMESPACE, "packagedElement");
        artifact.setAttribute("xmi:type", "uml:Artifact");
        String artifactId = XmiWriter.generateUUID();
        artifact.setAttribute("xmi:id", artifactId);
        artifact.setAttribute("name", artifactName);

        model.appendChild(artifact);
        return artifactId;
    }

    /**
     * Creates a UML Deployment relationship.
     */
    private void createDeployment(Document doc, Element model, String deploymentName,
            String nodeId, String artifactId) {
        Element deployment = doc.createElementNS(UML_NAMESPACE, "packagedElement");
        deployment.setAttribute("xmi:type", "uml:Deployment");
        deployment.setAttribute("xmi:id", XmiWriter.generateUUID());
        deployment.setAttribute("name", deploymentName);
        deployment.setAttribute("location", nodeId);
        deployment.setAttribute("deployedArtifact", artifactId);

        model.appendChild(deployment);
    }

    /**
     * Creates a UML CommunicationPath between two nodes.
     */
    private void createCommunicationPath(Document doc, Element model, String pathName,
            String sourceNodeId, String targetNodeId) {
        Element commPath = doc.createElementNS(UML_NAMESPACE, "packagedElement");
        commPath.setAttribute("xmi:type", "uml:CommunicationPath");
        commPath.setAttribute("xmi:id", XmiWriter.generateUUID());
        commPath.setAttribute("name", pathName);

        // Create member ends
        Element memberEnd1 = doc.createElementNS(UML_NAMESPACE, "memberEnd");
        memberEnd1.setAttribute("xmi:idref", sourceNodeId);
        commPath.appendChild(memberEnd1);

        Element memberEnd2 = doc.createElementNS(UML_NAMESPACE, "memberEnd");
        memberEnd2.setAttribute("xmi:idref", targetNodeId);
        commPath.appendChild(memberEnd2);

        model.appendChild(commPath);
    }

    /**
     * Finds the node where a service is deployed.
     */
    private String findNodeForService(String service, Map<String, Set<String>> nodeServices) {
        for (Map.Entry<String, Set<String>> entry : nodeServices.entrySet()) {
            if (entry.getValue().contains(service)) {
                return entry.getKey();
            }
        }
        return null;
    }

    /**
     * Extracts node/host information from metadata tags.
     * Supports multiple deployment platforms: Kubernetes, Docker, VMs, Cloud.
     */
    private String extractNode(Map<String, Object> metadata) {
        if (metadata == null)
            return "Node-Unknown";

        String nodeName = null;

        // Kubernetes: Try pod.name, k8s.pod.name
        Object podName = metadata.get("pod.name");
        if (podName == null) {
            podName = metadata.get("k8s.pod.name");
        }
        if (podName != null) {
            nodeName = podName.toString();
            // Extract base name (remove K8s hash suffix)
            nodeName = com.uml.generator.util.NameUtils.extractBaseName(nodeName);
            return nodeName;
        }

        // Docker: Try container.name, container.id
        Object containerName = metadata.get("container.name");
        if (containerName != null) {
            nodeName = containerName.toString();
            nodeName = com.uml.generator.util.NameUtils.extractBaseName(nodeName);
            return nodeName;
        }

        Object containerId = metadata.get("container.id");
        if (containerId != null) {
            String id = containerId.toString();
            // Use short form of container ID (first 12 chars)
            if (id.length() > 12) {
                id = id.substring(0, 12);
            }
            return "Container-" + id;
        }

        // VM/Cloud: Try hostname, instance.id, host.name, node.name
        Object hostname = metadata.get("hostname");
        if (hostname != null) {
            return hostname.toString();
        }

        Object instanceId = metadata.get("instance.id");
        if (instanceId != null) {
            return instanceId.toString();
        }

        Object hostName = metadata.get("host.name");
        if (hostName != null) {
            return hostName.toString();
        }

        Object nodeNameObj = metadata.get("node.name");
        if (nodeNameObj != null) {
            return nodeNameObj.toString();
        }

        Object hostIp = metadata.get("host.ip");
        if (hostIp != null) {
            return "Host-" + hostIp.toString();
        }

        // Generic fallback
        return "Node-" + Math.abs(metadata.hashCode() % 10000);
    }
}

File: python_version/jaeger_uml_generator/generators/deployment_diagram_generator.py
"""Deployment diagram generator."""

import logging
import xml.etree.ElementTree as ET
from typing import List, Dict, Set
from .diagram_generator import DiagramGenerator
from ..models import Trace
from ..analyzer import TraceAggregator
from ..renderer import XmiWriter, XmiFormat
from ..utils import extract_base_name


logger = logging.getLogger(__name__)


class DeploymentDiagramGenerator(DiagramGenerator):
    """Generates UML Deployment Diagrams in XMI 2.5.1 format from aggregated Jaeger traces."""
    
    def __init__(self, xmi_format: str = "papyrus"):
        """Initialize generator with XMI format.
        
        Args:
            xmi_format: Output format ('papyrus' or 'magicdraw')
        """
        format_enum = XmiFormat(xmi_format)
        self.xmi_writer = XmiWriter(format_enum)
    
    def get_diagram_type(self) -> str:
        return "deployment"
    
    def generate_xmi(self, traces: List[Trace]) -> str:
        """
        Generate XMI for deployment diagram from multiple traces.
        
        Args:
            traces: List of Trace objects
            
        Returns:
            XMI content as string
        """
        if not traces:
            logger.warning("No traces provided for deployment diagram generation")
            return ""
        
        try:
            aggregator = TraceAggregator(traces)
            
            # Determine model name
            model_name = "DeploymentDiagram"
            if traces and traces[0].source_name:
                model_name = f"{traces[0].source_name}_Deployment"
            
            # Create XMI document
            root = self.xmi_writer.create_xmi_document(model_name)
            model = self.xmi_writer.get_model_element(root)
            
            # Extract deployment information from service metadata
            service_metadata = aggregator.get_service_metadata()
            node_services: Dict[str, Set[str]] = {}
            node_ids: Dict[str, str] = {}
            
            # Analyze metadata to extract deployment info
            for service, metadata in service_metadata.items():
                node = self._extract_node(metadata)
                if node not in node_services:
                    node_services[node] = set()
                node_services[node].add(service)
            
            # Create nodes and deploy artifacts
            artifact_ids: Dict[str, str] = {}
            
            for node_name, services in node_services.items():
                # Create node
                node_elem, node_id = self.xmi_writer.create_packaged_element(
                    model, "Node", node_name
                )
                node_ids[node_name] = node_id
                
                # Create artifacts and deployments for services on this node
                for service in services:
                    artifact_elem, artifact_id = self.xmi_writer.create_packaged_element(
                        model, "Artifact", service
                    )
                    artifact_ids[service] = artifact_id
                    
                    self._create_deployment(
                        model,
                        f"{node_name}_deploys_{service}",
                        node_id,
                        artifact_id
                    )
            
            # Create communication paths between nodes
            dependencies = aggregator.get_service_dependencies()
            created_paths: Set[str] = set()
            
            for from_service, to_services in dependencies.items():
                # Find node for from_service
                from_node = self._find_node_for_service(from_service, node_services)
                
                for to_service in to_services:
                    to_node = self._find_node_for_service(to_service, node_services)
                    
                    if from_node and to_node and from_node != to_node:
                        path_key = f"{from_node}_to_{to_node}"
                        path_key_reverse = f"{to_node}_to_{from_node}"
                        
                        # Create bidirectional path only once
                        if path_key not in created_paths and path_key_reverse not in created_paths:
                            from_node_id = node_ids.get(from_node)
                            to_node_id = node_ids.get(to_node)
                            
                            if from_node_id and to_node_id:
                                self._create_communication_path(
                                    model,
                                    path_key,
                                    from_node_id,
                                    to_node_id
                                )
                                created_paths.add(path_key)
            
            total_nodes = len(node_services)
            total_services = len(service_metadata)
            
            logger.info(f"Generated deployment diagram XMI with {total_nodes} node(s) "
                       f"and {total_services} service(s)")
            
            return self.xmi_writer.document_to_string(root)
            
        except Exception as e:
            logger.error(f"Failed to generate deployment diagram XMI: {e}")
            return ""
    
    def _create_deployment(self, model: ET.Element, deployment_name: str,
                          node_id: str, artifact_id: str):
        """
        Create a UML Deployment relationship.
        
        Args:
            model: Parent model element
            deployment_name: Deployment name
            node_id: Node ID where artifact is deployed
            artifact_id: Artifact ID
        """
        self.xmi_writer.create_packaged_element(
            model, "Deployment", deployment_name,
            location=node_id, deployedArtifact=artifact_id
        )
    
    def _create_communication_path(self, model: ET.Element, path_name: str,
                                   source_node_id: str, target_node_id: str):
        """
        Create a UML CommunicationPath between two nodes.
        
        Args:
            model: Parent model element
            path_name: Path name
            source_node_id: Source node ID
            target_node_id: Target node ID
        """
        comm_path, _ = self.xmi_writer.create_packaged_element(
            model, "CommunicationPath", path_name
        )
        
        # Create member ends (using ownedEnd instead of memberEnd for better compatibility)
        end1 = ET.SubElement(comm_path, "ownedEnd")
        end1.set(f"{{{self.xmi_writer.XMI_NAMESPACE}}}id", self.xmi_writer.generate_uuid())
        end1.set("type", source_node_id)
        
        end2 = ET.SubElement(comm_path, "ownedEnd")
        end2.set(f"{{{self.xmi_writer.XMI_NAMESPACE}}}id", self.xmi_writer.generate_uuid())
        end2.set("type", target_node_id)
    
    def _find_node_for_service(self, service: str, 
                               node_services: Dict[str, Set[str]]) -> str:
        """
        Find the node where a service is deployed.
        
        Args:
            service: Service name
            node_services: Map of node names to services
            
        Returns:
            Node name or None
        """
        for node_name, services in node_services.items():
            if service in services:
                return node_name
        return None
    
    def _extract_node(self, metadata: Dict[str, any]) -> str:
        """
        Extract node/host information from metadata tags.
        Supports multiple deployment platforms: Kubernetes, Docker, VMs, Cloud.
        
        Args:
            metadata: Service metadata dictionary
            
        Returns:
            Node name
        """
        if not metadata:
            return "Node-Unknown"
        
        # Kubernetes: Try pod.name, k8s.pod.name
        pod_name = metadata.get('pod.name') or metadata.get('k8s.pod.name')
        if pod_name:
            # Extract base name (remove K8s hash suffix)
            return extract_base_name(str(pod_name))
        
        # Docker: Try container.name, container.id
        container_name = metadata.get('container.name')
        if container_name:
            return extract_base_name(str(container_name))
        
        container_id = metadata.get('container.id')
        if container_id:
            # Use short form of container ID (first 12 chars)
            container_id_str = str(container_id)
            if len(container_id_str) > 12:
                container_id_str = container_id_str[:12]
            return f"Container-{container_id_str}"
        
        # VM/Cloud: Try hostname, instance.id, host.name, node.name
        hostname = metadata.get('hostname')
        if hostname:
            return str(hostname)
        
        instance_id = metadata.get('instance.id')
        if instance_id:
            return str(instance_id)
        
        host_name = metadata.get('host.name')
        if host_name:
            return str(host_name)
        
        node_name = metadata.get('node.name')
        if node_name:
            return str(node_name)
        
        host_ip = metadata.get('host.ip')
        if host_ip:
            return f"Host-{host_ip}"
        
        # Generic fallback
        return f"Node-{abs(hash(str(metadata)) % 10000)}"
